lib/spack/spack/solver/asp.py

```diff
@@ -2433,14 +2446,23 @@ def visit(node):
         path = os.path.join(parent_dir, "concretize.lp")
         parse_files([path], visit)
 
+    def add_runtime_virtuals(self):
+        """Explicitly add the virtuals that might be attached dynamically by runtimes"""
+        for vpkg in ("fortran-rt", "ifcore", "gfortran"):
```

Comment:
I can probably iterate over packages tagged as "runtimes", and check which virtuals they could provide. Let me try that.

Reply:
Why aren't these added based on what is passed to `depends_on` in `RuntimePropertyRecorder`?

You know that someone executed `runtime_constraints` and added a possible dependency on these virtuals, and you know that they're virtual because they're provided by the runtime packages.  This shouldn't need to be hard-coded...

Reply:
I'll check if it is easy to unify that. This was added here to follow the same code path for max duplicate counters as for everything else, but it makes sense to aggregate the way you say.

Reply:
I removed this in https://github.com/spack/spack/pull/42062/commits/b9c497abec6db0f0665fc7f6a454823bc961bf7b and moved everything to the counters

Reply:
I don't think moving this to the counters really solves the issue.  That replaces the list with tags that users have to add.  See my comment below (and the one above) -- we should just *know* what possible virtuals there are before solve.  You do not know them from `possible_dependencies()`, but you *do* know them once you've gone through those dependencies, figured out which ones have a `runtime_constraints()` method, and called it for each of them.